File: example/src/main/java/ru/virgil/example/system/BaseEntity.java
package ru.virgil.example.system;

import lombok.Data;
import lombok.experimental.ExtensionMethod;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import javax.persistence.MappedSuperclass;
import java.time.LocalDateTime;

@MappedSuperclass
@Data
@ExtensionMethod(value = {LocalDateTime.class, DateApiExtensions.class}, suppressBaseMethods = false)
public abstract class BaseEntity {

    @CreationTimestamp
    private LocalDateTime createdAt;
    @UpdateTimestamp
    private LocalDateTime updatedAt;

    public LocalDateTime getCreatedAt() {
        return createdAt.databaseTruncate();
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt.databaseTruncate();
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt.databaseTruncate();
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt.databaseTruncate();
    }
}

File: example/src/main/java/ru/virgil/example/system/entity/BaseEntity.kt
package ru.virgil.example.system.entity

import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import ru.virgil.example.system.entity.LocalDateTimeExtensions.databaseTruncate
import java.time.LocalDateTime
import javax.persistence.MappedSuperclass

@MappedSuperclass
abstract class BaseEntity(
    createdAt: LocalDateTime = LocalDateTime.now(),
    updatedAt: LocalDateTime = LocalDateTime.now(),
    var deleted: Boolean = false,
) {

    @UpdateTimestamp
    var updatedAt = updatedAt
        get() = field.databaseTruncate()
        set(value) {
            field = value.databaseTruncate()
        }

    @CreationTimestamp
    var createdAt = createdAt
        get() = field.databaseTruncate()
        set(value) {
            field = value.databaseTruncate()
        }
}
